### DISCUSSION TITLE
break up the code

### DISCUSSION THREAD

**Discussion by Data-Monkey** [netalertx/NetAlertX#236] [General]
Hi, I have started to break up the code.

https://github.com/Data-Monkey/Pi.Alert/tree/split_it_up

it has lots of global variables and db connections everywhere.
Still lots more work to be done.

Do you have some tests that you are running?

**Comment by Data-Monkey** [comment]
Hi,
I think I have the code broken out a lot now. Maybe too much for my own good.
It is full of global varaibles and that is a pain to split, but I have a workaround via config module.
Now I need to test it. 
How do you generally test this thing?
I saw your docker compose file. Do you just break out all the relevant files from the container to your code and run it like that? 
Do you have a set of test-cases? 
Thanks

**Reply by jokob-sk** [reply]
Thanks a lot!

I don't have any specific test cases, but what I usually test is:

* Blank setup with no DB or config
* existing DB / config
* sending a notification (e. g. Delete a device and wait for a scan to run)
* saving settings 
* Notification gateways tested:
  * email, Apprise, Telegram (webhook) 
* test a couple of plugins
* check the Error log for anything unusual 

I know that is not very structured, but I do what I can :) also I'm super busy with work right now so I hope I'll have energy to look at this over the weekend. 

If the above works feel free to open a PR and I might be able to merge it over the weekend. 

Let me know if you have any questions!

**Reply by Data-Monkey** [reply]
thanks, that is a good start. 
No, I don't think it will be ready for a PR this weekend. 
And if you need to run some other fixes, please don't wait for this. I will try to catch up to your version before I make a PR.
If you have time, please have a look at my branch and see if this is going to far ...

**Reply by jokob-sk** [reply]
Thanks so much for such a thorough approach, really appreciate it! 
Happy to wait, this kind of maintenance is much more important than some arbitrary release frequency :)

**Reply by jokob-sk** [reply]
I had a quick look at your branch and it looks solid.

When you said you want to split it up I thought maybe like 5 files, but this is so much more than I expected. I saw some methods which I'd place logically into a different file, but these things can be ironed out later. 

Hope you find a way how to solve the circular references which I saw you mentioned in a couple of places. 

Also, thanks for unifying the event log messages :) This will be a significant update.

**Reply by Data-Monkey** [reply]
Thanks, I might have taken it a bit too far :-)

> I saw some methods which I'd place logically into a different file,

please let me know what these are, any feedback is welcome.
I am still struggling to see the whole picture of this project.

> Hope you find a way how to solve the circular references
I might need more files to solve these ... I guess that is the benefit of one large file, everything can see everything else.

Can I ask you, why are the plugins in the /pialert/**front**/ folder?

**Reply by jokob-sk** [reply]
> Can I ask you, why are the plugins in the /pialert/front/ folder?

As it generates a front end I thought it's logical but it can be moved if you think it's a better idea to move it somewhere - links in the app and docs will have to be updated though.

**Reply by jokob-sk** [reply]
Just remembered something - I was considering allowing to add custom UI components for specific plugins, that's why I placed it under /front

**Comment by Data-Monkey** [comment]
Hi Jakob, I have done some more cleanup but am getting to the limit without breaking down things into smaller methods.
I am surprised that the different scans are all done so differently. 
The ArpScan writes to a variable where all the others write to their own tables. ??
Then there is someplace where the individual tables get "merged" into the CurrentScan table and from there into Devices. 
I would love to make it all more consistent but I am afraid I have not understood some of the things happening. Especially with the parallelism in the code. Also, what is cycle doing? It seems to be used for lots of things and I don't understand it ...

Timezone also seem to be used inconsistently and did not work for me. So I removed all the TZ from the docker build and just pass in the correct timezone, that seems to work for me, but there was a lot of code in the files, so I assume it did not work for you previously...

For testing I am trying a normal scan, nmap scan, pihole network and pihole dhcp. 
I have not tried the other scans and I have not tried the notifications. There is bound to be something wrong in there...

**Comment by jokob-sk** [comment]
Hi,

I'll try answering what I know.

Arpscan was the the only scan that was available in the original code. I wanted to make it more extensible by creating a separate nmap scan table. I realised that doing this for every scan is unsustainable so I came up with the plugin system. The Nmap scan could be rewritten into a plugin later on.

Cycle - no idea now. Tried to remove it but it's used in way too many sql queries. Was used in the old code to change the frequency of the scan when the execution was based on a cron script.

TZ was also previously used for adjusting the cron script so if it can be used without then great. Although ideally existing docker compose set ups should work.

It would be great to test a couple of the plugins and the notifications, but I could do that probably over the next weekend.

Thanks!

**Reply by Data-Monkey** [reply]
> Arpscan was the the only scan that was available in the original code. I wanted to make it more extensible by creating a separate nmap scan table. I realised that doing this for every scan is unsustainable so I came up with the plugin system. The Nmap scan could be rewritten into a plugin later on.

That sounds good. I think all the different scans can be treated the same and update the same table. Then the rest of the script becomes a lot easier to manage. Let me think about that. 

Do you know how many installations of your Pi.Alert are out there??

**Reply by jokob-sk** [reply]
Re number of installs:

I'd be guessing around 10k if I was optimistic. I get around 3k pulls in 24h which I assume could be watchtower checks? Not sure.

But yeah, backward compatibility is pretty important to me if that's why you asked.

**Reply by Data-Monkey** [reply]
> I'd be guessing around 10k if I was optimistic.

That is cool! I will try to test it with your original docker-compose file to see what happens. 
Since it is only supporting to run in a docker, I would suggest we remove the TZ configuration from the frontend and config file and make sure the container runs with the appropriate time zone. 
But that is probably not as important as making it all work again. I wen't a bit too far taking the file appart ...

**Reply by jokob-sk** [reply]
Yep definitely I vote for incremental improvements.

**Reply by Data-Monkey** [reply]
Do you think we can have some kind of Beta release to get more people to test it and to not break all the other peoples confidence. 
The guys that go for beta usually are aware what they are doing might have some impact.

**Reply by jokob-sk** [reply]
Yep, thinking the same. Pre realease candidate or something in those lines, although not sure how the ci/CD via github actions would handle that. Will have to check it over the weekend.

If you can pull down some of the changes from my branch and submit a PR so I can try to merge it.

**Reply by Data-Monkey** [reply]
maybe keep the split code in a seperate branch for now until it is stable enough. 
I will have a look at the changes in your master and merge/retrofit. I think most changes were in the frontend and I did not touch that.

**Reply by jokob-sk** [reply]
Sure here you go: https://github.com/jokob-sk/Pi.Alert/tree/split-up-work-2023-05-30

**Comment by jokob-sk** [comment]
Alright, your changes are merged (hopefully correctly) into this branch:

https://github.com/jokob-sk/Pi.Alert/tree/split-up-work-2023-05-30

There is also a dev image to test: 

https://registry.hub.docker.com/r/jokobsk/pi.alert_dev

- I'll be testing this now over the weekend and if nothing comes up will merge it (or make this the main branch) towards the end of the weekend. (not sure if a pre-release is a good idea)

- I'll then leave it be and aim for a release (or pre-release) next weekend. 

Thanks again for all the work! @Data-Monkey 

BTW - try pulling the default DB file into your branch, there is still some other DB file present.

**Reply by jokob-sk** [reply]
Issue 1: 

In code

``` python
def mqtt_start(db):    

    #global client

    if conf.mqtt_connected_to_broker == False:
        conf.mqtt_connected_to_broker = True           
        client = mqtt_create_client() 
    
    # General stats    

    # Create a generic device for overal stats
    create_generic_device(client)

    # Get the data
    row = get_device_stats(db)   

    columns = ["online","down","all","archived","new","unknown"]

    payload = ""

    # Update the values 
    for column in columns:       
        payload += '"'+column+'": ' + str(row[column]) +','  
```

The following error is thrown when an MQTT notification is raised:

```bash

pialert    | Traceback (most recent call last):
pialert    |   File "/usr/lib/python3.9/runpy.py", line 197, in _run_module_as_main
pialert    |     return _run_code(code, main_globals, None,
pialert    |   File "/usr/lib/python3.9/runpy.py", line 87, in _run_code
pialert    |     exec(code, run_globals)
pialert    |   File "/home/pi/pialert/pialert/__main__.py", line 314, in <module>
pialert    |     sys.exit(main())
pialert    |   File "/home/pi/pialert/pialert/__main__.py", line 277, in main
pialert    |     send_notifications(db)
pialert    |   File "/home/pi/pialert/pialert/reporting.py", line 307, in send_notifications
pialert    |     mqtt_start(db)
pialert    |   File "/home/pi/pialert/pialert/publishers/mqtt.py", line 165, in mqtt_start
pialert    |     payload += '"'+column+'": ' + str(row[column]) +','
pialert    | TypeError: list indices must be integers or slices, not str

```

Seems like:

```python
#-------------------------------------------------------------------------------
def get_device_stats(db):
    # columns = ["online","down","all","archived","new","unknown"]
    return db.read(sql_devices_stats)
```

returns an incorrect DataType.

I might be wrong though... havent't used `db.read()` before.

⚠Please also double check if  `def get_all_devices(db):  ` returns the correct type as part of this fix.

**Reply by jokob-sk** [reply]
Issue 2:

In code:

```python
        if conf.REPORT_WEBHOOK and check_config('webhook'):
            updateState(db,"Send: Webhook")
            mylog('info', ['[Notification] Sending report by Webhook'])
            send_webhook (json_final, mail_text)
```

The following error is thrown:

```bash
pialert    | Traceback (most recent call last):
pialert    |   File "/usr/lib/python3.9/runpy.py", line 197, in _run_module_as_main
pialert    |     return _run_code(code, main_globals, None,
pialert    |   File "/usr/lib/python3.9/runpy.py", line 87, in _run_code
pialert    |     exec(code, run_globals)
pialert    |   File "/home/pi/pialert/pialert/__main__.py", line 314, in <module>
pialert    |     sys.exit(main())
pialert    |   File "/home/pi/pialert/pialert/__main__.py", line 277, in main
pialert    |     send_webhook (json_final, mail_text)
pialert    | TypeError: send() takes 1 positional argument but 2 were given

```

**Reply by jokob-sk** [reply]
Issue 3:


The config file is imported every loop (every 5 seconds), because the `lastTimeImported = 0`:

https://github.com/jokob-sk/Pi.Alert/blob/de62956c4010763c4dfe82a0c1aa7696c56856ab/pialert/initialise.py#L46

and then:

https://github.com/jokob-sk/Pi.Alert/blob/de62956c4010763c4dfe82a0c1aa7696c56856ab/pialert/initialise.py#L52

```bash
pialert    | 11:32:06 [Config] Imported new config
```
Probably this is how it should be used:

```python
    # Used to determine the next import
    conf.lastTimeImported = time.time()
```

**Reply by jokob-sk** [reply]
Issue 4:

(so it's easier for you to track these)

Default DB file not used

**Reply by jokob-sk** [reply]
Issue 5:

I'm getting these periodic dropouts in the online history table:

![image](https://github.com/jokob-sk/Pi.Alert/assets/96159884/211db928-a11b-4550-a6b6-4abc5134bf98)

![image](https://github.com/jokob-sk/Pi.Alert/assets/96159884/1c41608d-d1f5-4c46-90a6-f2f12cd2b38d)


This might be related to DB locks, but I'm only guessing...

The data should be saved to the DB in this method AFAIK:

https://github.com/Data-Monkey/Pi.Alert/blob/b814a856d0ff4b6c148c844b205528e4246ac3c9/pialert/database.py#LL417C20-L417C20

**Reply by jokob-sk** [reply]
Issue 6:

(optional)

The README is not finished, but I'm happy to finish it once we solve the code related issues:

https://github.com/Data-Monkey/Pi.Alert/blob/b814a856d0ff4b6c148c844b205528e4246ac3c9/pialert/README.md

**Reply by Data-Monkey** [reply]
> Issue 3:
> 
> The config file is imported every loop (every 5 seconds), because the `lastTimeImported = 0`:
> 
> https://github.com/jokob-sk/Pi.Alert/blob/de62956c4010763c4dfe82a0c1aa7696c56856ab/pialert/initialise.py#L46
> 
> and then:
> 
> https://github.com/jokob-sk/Pi.Alert/blob/de62956c4010763c4dfe82a0c1aa7696c56856ab/pialert/initialise.py#L52
> 
> ```shell
> pialert    | 11:32:06 [Config] Imported new config
> ```
> 
> Probably this is how it should be used:
> 
> ```python
>     # Used to determine the next import
>     conf.lastTimeImported = time.time()
> ```

I changed it to compare modified times between the runs. This avoids TimeZone issues.
Fixed in this Commit: https://github.com/Data-Monkey/Pi.Alert/commit/6bb891f830fc03714b98ee5d594aa11c8fa32303

```python
def importConfigs (db): 

    sql = db.sql

    # get config file name
    config_file = Path(fullConfPath)

    # Only import file if the file was modifed since last import.
    # this avoids time zone issues as we just compare the previous timestamp to the current time stamp
    mylog('debug', ['[Import Config] checking config file '])
    mylog('debug', ['[Import Config] lastImportedConfFile     :', conf.lastImportedConfFile])
    mylog('debug', ['[Import Config] file modified time       :', os.path.getmtime(config_file)])
    

    if (os.path.getmtime(config_file) == conf.lastImportedConfFile) :
        mylog('debug', ['[Import Config] skipping config file import'])
        return

    conf.lastImportedConfFile = os.path.getmtime(config_file)  
```

**Reply by Data-Monkey** [reply]
> Issue 4:
> 
> (so it's easier for you to track these)
> 
> Default DB file not used

Added Default DB back from your main repo

**Reply by Data-Monkey** [reply]
**Issue 2** 
> ```python
> send_webhook (json_final, mail_text)
> ```

fixed 
 ```python
send_webhook (msg)
```

**Reply by jokob-sk** [reply]
Merged your fixes for Issues 2,3,4, thanks for that!

Issue 2 > still getting an error when sending a webhook. (see below)
Issue 3 > resolved
Issue 4 > Should be resolved
Issue 5 > NOT resolved:

![image](https://github.com/jokob-sk/Pi.Alert/assets/96159884/88bbc713-4f71-4800-a56d-1b72885b563b)

Edit: Spoke too soon, issue 5 is only not occurring if container crashes after trying to send a webhook. 

![image](https://github.com/jokob-sk/Pi.Alert/assets/96159884/1a5b8036-d90c-4554-92b9-9e47077641f5)

**Reply by jokob-sk** [reply]
Issue 7:

Webhook still not being sent.

``` bash
pialert    | Traceback (most recent call last):
pialert    |   File "/usr/lib/python3.9/runpy.py", line 197, in _run_module_as_main
pialert    |     return _run_code(code, main_globals, None,
pialert    |   File "/usr/lib/python3.9/runpy.py", line 87, in _run_code
pialert    |     exec(code, run_globals)
pialert    |   File "/home/pi/pialert/pialert/__main__.py", line 314, in <module>
pialert    |     sys.exit(main())
pialert    |   File "/home/pi/pialert/pialert/__main__.py", line 276, in main
pialert    |     send_notifications(db)
pialert    |   File "/home/pi/pialert/pialert/reporting.py", line 288, in send_notifications
pialert    |     send_webhook (msg)
pialert    |   File "/home/pi/pialert/pialert/publishers/webhook.py", line 54, in send
pialert    |     mylog('debug', '[send_webhook] curlParams: '+  curlParams)
pialert    | TypeError: can only concatenate str (not "list") to str

```

`+` should be probably changed to  `,`

**Reply by Data-Monkey** [reply]
Hi thanks for testing, I created these now as issues in my repo, which should make it even easier to track. Please feel free to add more.
And don't worry about the template, I guess that is your template, but we don't need that for these issues.

**Comment by Data-Monkey** [comment]
Thanks for merging all my code back into your project. That was a brave move! 
Let me know where you want to go from here.

**Comment by jokob-sk** [comment]
Thank you for spending so much time on this!
I'll try to probably release it over the weekend.

Where next? I don't have any specific plan. Usually there is a feature request that inspires me to implement some broader functionality. Now I'm mostly thinking about how to enable the community to contribute, such as via the plugins or detailed docs.

Creating notification plugin system and/or rewriting the Nmap scan into a plugin could be next logical steps. Open to suggestions though. 😉

**Comment by Data-Monkey** [comment]
some good ideas!
I was looking at trying to align to the PEP8 formatting of the code. Again, that does not add any functionality but should make it more readable / maintainable. 

I like the plugin idea and maybe also dividing the "scanners" into different types of scanners. 
At the moment some are purely there to discover devices but don't provide the online/offline status. Others do the status and others again try to dicover more details. Maybe they can be split and run at seperate times. 
And plugins could have a type so they get executed at the right time.
A plugin system to add more notification mechanisms would be nice as well.

But I also expect there will be many bugs to iron out with the split up version. I certainly did not manage to run all the code yet.

**Comment by jokob-sk** [comment]
Yep, agree on possible improvements.

You probably know how it is, I also try to balance personal life lol, so sometimes I have bandwidth for sweeping changes and sometimes for high priority maintenance tasks only.

Re bugs, I was running the build for a week and hope I've tested most scenarios, but no bugs no progress and I hope the users will be understanding 👍

**Reply by Data-Monkey** [reply]
absolutely understand the priority for personal life balance!

**Comment by Data-Monkey** [comment]
Hi Jakob, 
sorry I let you hang there after I split the code and you merged it in so quickly. Life got busy and also my home-lab decided to give up after over 8 years of running without fail. 
I tried to revive a little laptop to run the docker but I think I need to fix the lab first. Haven't done any hardware selection in many years, it has all changed. 
Hopefully I can jump back in soon and help cleaning up some of the mess I made. 
Sorry!

**Comment by jokob-sk** [comment]
Thanks @Data-Monkey!
Not at all, no mess made, only improved 😊 (also I did a concious decision to merge and release it and I have to say it's not too bad 😅) 

I appreciate the follow up and looking forward to help 😁

My homeland is a Synology and a NUC, but 🤞 they won't die on me, but I feel your pain.

Would most likely build something custom though after 2 years with the Synology.

I also understand that IRL sometimes doesn't play nice with hobby projects and FOSS initiatives.

Looking forward to the next contribution, but no rush!
J
